service/history/workflow/update/update.go

```diff
@@ -156,34 +172,100 @@ func newCompleted(
 	return upd
 }
 
-// WaitOutcome observes this Update's completion, returning the Outcome when it
-// is available. This call will block until the Outcome is known or the provided
-// context.Context expires. It is safe to call this method outside of workflow lock.
-func (u *Update) WaitOutcome(ctx context.Context) (*updatepb.Outcome, error) {
-	return u.outcome.Get(ctx)
+// WaitLifecycleStage waits until the Update has reached at least `waitStage` or
+// a timeout. If the Update reaches `waitStage` with no timeout, the most
+// advanced stage known to have been reached is returned, along with the outcome
+// if any. If there is a timeout due to the supplied soft timeout, then
+// unspecified stage and nil outcome are returned, without an error. If there is
+// a timeout due to context deadline expiry, then the error is returned as usual.
+func (u *Update) WaitLifecycleStage(
+	ctx context.Context,
+	waitStage enumspb.UpdateWorkflowExecutionLifecycleStage,
+	softTimeout time.Duration) (UpdateStatus, error) {
+
+	switch waitStage {
+	case acceptedStage:
+		return u.waitLifecycleStage(ctx, u.WaitAccepted, softTimeout)
+	case completedStage:
+		return u.waitLifecycleStage(ctx, u.WaitOutcome, softTimeout)
+	default:
+		err := serviceerror.NewUnimplemented(fmt.Sprintf("%v is not implemented", waitStage))
```

Comment:
You used InvalidArgument for the same type of error above.
We should be consistent here.
The implication of Unimplemented is that it counts against our SLA compared to InvalidArgument which is a client issue.

Suggestion: even if you we don't typically use this form in the codebase, I think it'd be nicer to use `status.Errorf(codes.Unimplemented, "%v is not implemented", waitStage)`

Reply:
Good catch, I thought I'd changed all of them to `InvalidArgument`. Fixed.

I don't have the knowledge to adjudicate on `serviceerror.InvalidArgument` vs `status.Errorf` yet so I vote that we leave that for a more systemic replacement.